### ISSUE TITLE
overriding `make_train_function` does not work.

### ISSUE THREAD

**Issue by mmalekzadeh** [tensorflow/tensorflow#38797]
**System information**
- TensorFlow version (you are using): 2.2.X
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
Reference: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/training.py

Overriding the `train_step`  function of the Model class works well. But if you want to go a bit deeper down and override ` make_train_function` then it does not work.

In the document, it says that it is possible

`This method can be overridden to support custom training logic.`

But when you do that, it gives some errors. I tried to import what it asks for in the errors but it still wants more to import. 

Here is a sample code based on the recent François Chollet's notebook:

https://colab.research.google.com/drive/1pQX2pjXSU1AU182LuQge7xNeBJcFA3yb

Here you see that when you override it, it gives an error.

**Will this change the current api? How?**
It can.

**Who will benefit with this feature?**
Everybody who wants to write a customized training loop while taking advantage of the Keras `fit`.

As François Chollet' says:

"what if you need a custom training algorithm, but you still want to benefit from the convenient features of fit(), such as callbacks, built-in distribution support, or step fusing?"

**Comment by omalleyt12** [comment]
@mmalekzadeh Thanks for the issue!

Looks like there's two things going on in the example provided:

(1) The code is using code copy & pasted from `Model.make_train_function`. The default `Model.make_train_function` uses internal private TF APIs. To override, you will have to use public APIs

(2) The code is using the version at HEAD, which contains code that doesn't exist in 2.2 

In general, I'd recommend only overriding `Model.make_train_function` if you want very low-level control over the `tf.function` that gets created and used in `Model.fit`. This should only be used for very advanced use cases. Overriding `Model.train_step` should be enough for most use cases. If `Model.make_train_function` is overridden, you will have to return a `tf.function` that accepts an `tf.data.Iterator` and returns a dict of logs, and should construct this `tf.function` using public APIs available in TF2.2

**Comment by mmalekzadeh** [comment]
@omalleyt12  Many thanks for your response.

I have tried to use public API, but still couldn't manage to build a simple example of overriding `make_train_function`. I agree that `This should only be used for very advanced use cases` and this is what I really wanted to do. 

Btw, It seems the only option, for now, is to manage whatever we need to do in the `Model.train_step` and forget to override `Model.make_train_function` at all.
